diff --git a/CHANGES b/CHANGES
index bcd1e1c..b20efbb 100644
--- a/CHANGES
+++ b/CHANGES
@@ -281,7 +281,7 @@ CHANGES FROM 3.4 TO 3.5
   if selecting the element below the deleted one fails (because as the last
   one), select the one above it instead.
 
-* Add --enable-jemalloc to build with jemalloc memory allocator (since glibc
+* Add --enable-mimalloc to build with mimalloc memory allocator (since glibc
   malloc is so poor).
 
 * Add a way (refresh-client -r) for control mode clients to provide OSC 10 and
diff --git a/configure.ac b/configure.ac
index fa47482..5ee76d1 100644
--- a/configure.ac
+++ b/configure.ac
@@ -558,21 +558,21 @@ if test "x$found_malloc_trim" = xyes; then
 	AC_DEFINE(HAVE_MALLOC_TRIM)
 fi
 
-# Build against jemalloc if requested.
+# Build against mimalloc if requested.
 AC_ARG_ENABLE(
-	jemalloc,
-	AS_HELP_STRING(--enable-jemalloc, use jemalloc if it is installed)
+	mimalloc,,
+	AS_HELP_STRING(--enable-mimalloc, use mimalloc if it is installed)
 )
-if test "x$enable_jemalloc" = xyes; then
+if test "x$enable_mimalloc" = xyes; then
 	PKG_CHECK_MODULES(
-		JEMALLOC,
-		jemalloc,
+		MIMALLOC,
+		mimalloc,
 		[
-			AM_CPPFLAGS="$JEMALLOC_CFLAGS $AM_CPPFLAGS"
+			AM_CPPFLAGS="$MIMALLOC_CFLAGS $AM_CPPFLAGS"
 			CPPFLAGS="$AM_CPPFLAGS $SAVED_CPPFLAGS"
-			LIBS="$LIBS $JEMALLOC_LIBS"
+			LIBS="$LIBS $MIMALLOC_LIBS"
 		],
-		AC_MSG_ERROR("jemalloc not found")
+		LIBS="$LIBS -lmimalloc"
 	)
 fi
 

