not migrate too often
diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index 704d0af8e..9bea29ed0 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -12933,6 +12933,7 @@ static int sched_balance_newidle(struct rq *this_rq, struct rq_flags *rf)
 		update_next_balance(sd, &next_balance);
 
 		if (this_rq->avg_idle < curr_cost + sd->max_newidle_lb_cost)
+		if (this_rq->avg_idle/2 < curr_cost + sd->max_newidle_lb_cost)
 			break;
 
 		if (sd->flags & SD_BALANCE_NEWIDLE) {

