diff --git a/Makefile b/Makefile
index 3060ef5f0..eebebc925 100644
--- a/Makefile
+++ b/Makefile
@@ -872,6 +872,13 @@ KBUILD_RUSTFLAGS += -Copt-level=2
 else ifdef CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE_O3
 KBUILD_CFLAGS += -O3
 KBUILD_CFLAGS += $(call cc-option,-mllvm -enable-pipeliner)
+KBUILD_CFLAGS += $(call cc-option,-mllvm -extra-vectorizer-passes)
+KBUILD_CFLAGS += $(call cc-option,-mllvm -enable-cond-stores-vec)
+KBUILD_CFLAGS += $(call cc-option,-mllvm -aggressive-ext-opt)
+KBUILD_CFLAGS += $(call cc-option,-mllvm -enable-gvn-hoist)
+KBUILD_CFLAGS += $(call cc-option,-mllvm -enable-gvn-sink)
+KBUILD_CFLAGS += $(call cc-option,-mllvm -enable-loop-distribute)
+KBUILD_CFLAGS += $(call cc-option,-mllvm -slp-vectorize-hor-store)
 KBUILD_RUSTFLAGS += -Copt-level=3
 else ifdef CONFIG_CC_OPTIMIZE_FOR_SIZE
 KBUILD_CFLAGS += -Os

