diff --git a/mm/memcontrol.c b/mm/memcontrol.c
index 86f43b7e5..b77199337 100644
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -773,7 +773,7 @@ void mod_lruvec_state(struct lruvec *lruvec, enum node_stat_item idx,
 	mod_node_page_state(lruvec_pgdat(lruvec), idx, val);
 
 	/* Update memcg and lruvec */
-	if (!mem_cgroup_disabled())
+	if (likely(!mem_cgroup_disabled()))
 		mod_memcg_lruvec_state(lruvec, idx, val);
 }
 

