diff --git a/config.c b/config.c
index 14e836c1..61fe5c60 100644
--- a/config.c
+++ b/config.c
@@ -1604,8 +1604,8 @@ parse_section_colors_light(struct context *ctx)
 static bool
 parse_section_colors(struct context *ctx)
 {
-    LOG_WARN("%s:%d: [colors]: deprecated; use [colors-dark] instead",
-             ctx->path, ctx->lineno);
+    // LOG_WARN("%s:%d: [colors]: deprecated; use [colors-dark] instead",
+    //          ctx->path, ctx->lineno);
 
     user_notification_add(
         &ctx->conf->notifications,

