diff --git a/terminal.h b/terminal.h
index 364d57b3..bd37806f 100644
--- a/terminal.h
+++ b/terminal.h
@@ -163,9 +163,9 @@ struct row {
 
     struct {
         bool prompt_marker;
-        int cmd_start;  /* Column, -1 if unset */
-        int cmd_end;    /* Column, -1 if unset */
-    } shell_integration;
+        int16_t cmd_start;  /* Column, -1 if unset */
+        int16_t  cmd_end;    /* Column, -1 if unset */
+    } __attribute__((packed)) shell_integration;
 };
 
 struct sixel {
