diff --git a/malloc/malloc.c b/malloc/malloc.c
index 27dfd1eb..89838a6a 100644
--- a/malloc/malloc.c
+++ b/malloc/malloc.c
@@ -310,7 +310,7 @@
 
 /* This is another arbitrary limit, which tunables can change.  Each
    tcache bin will hold at most this number of chunks.  */
-# define TCACHE_FILL_COUNT 7
+# define TCACHE_FILL_COUNT 24
 
 /* Maximum chunks in tcache bins for tunables.  This value must fit the range
    of tcache->counts[] entries, else they may overflow.  */

